File: swift/src/main/java/androidx/swift/session/SessionManager.java
/*
 * Copyright 2022 RAE
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.swift.session;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class SessionManager {

    @Nullable
    protected static SessionManager sSessionManager;

    public static class Builder {

        private final Context mContext;
        private String mSessionName;
        private Class<?> mUserClass;
        private boolean mIsJsonMode;

        public Builder(Context context) {
            mContext = context.getApplicationContext();
        }

        public Builder setSessionName(String name) {
            mSessionName = name;
            return this;
        }

        public Builder setUserClass(Class<?> cls) {
            mUserClass = cls;
            return this;
        }

        public Builder setJsonMode(boolean isJsonMode) {
            mIsJsonMode = isJsonMode;
            return this;
        }

        public SessionManager build() {
            if (mUserClass == null) {
                throw new NullPointerException("session user class can't null");
            }
            if (TextUtils.isEmpty(mSessionName)) {
                mSessionName = mContext.getPackageName() + ".session";
            }
            if (mIsJsonMode) {
                return new JsonFileSessionManager(mContext, mSessionName, mUserClass);
            }
            return new SharedPreferencesSessionManager(mContext, mSessionName, mUserClass);
        }
    }

    /**
     * 初始化默认的用户信息
     * @param context Context
     * @param userInfoClass 用户信息类
     * @deprecated 请使用 {@link Builder} 来初始化，调用{@link #setSessionManager(SessionManager)} 设置默认会话管理器
     */
    @Deprecated
    public static void initDefaultSessionManager(Context context, Class<?> userInfoClass) {
        SessionManager sessionManager = new Builder(context).setUserClass(userInfoClass).build();
        setSessionManager(sessionManager);
    }

    public static void init(Builder builder) {
        setSessionManager(builder.build());
    }

    /**
     * 设置默认的会话管理器
     * @param sessionManager 会话管理器
     */
    public static void setSessionManager(SessionManager sessionManager) {
        if (sSessionManager == null) {
            synchronized (SessionManager.class) {
                if (sSessionManager == null) {
                    sSessionManager = sessionManager;
                }
            }
        }
    }

    /**
     * 获取默认的会话管理器
     */
    @NonNull
    public static SessionManager getDefault() {
        if (sSessionManager == null) {
            throw new NullPointerException("default session manager is null.");
        }
        return sSessionManager;
    }


    @Nullable
    private List<SessionStateListener> mSessionStateListeners;
    private final String mSessionName;

    protected SessionManager(String sessionName) {
        this.mSessionName = sessionName;
    }

    /**
     * 获取当前会话管理器名称
     * @return 名称
     */
    public String getSessionName() {
        return mSessionName;
    }

    /**
     * 是否登录
     */
    public boolean isLogin() {
        return getUserInfo() != null;
    }


    /**
     * 清除会话信息，即退出登录。
     */
    public abstract void forgot();

    /**
     * 获取当前登录的用户信息，在调用该方法之前请先调用{@link #isLogin()}来判断是否登录
     * @return 用户信息
     */
    @Nullable
    public abstract <T> T getUserInfo();

    /**
     * 设置用户信息
     * @param userInfo 用户信息
     * @param <T> 用户实体
     */
    protected abstract <T> void onSaveUserInfo(T userInfo);

    /**
     * 设置当前用户信息
     */
    public <T> void setUserInfo(T user) {
        this.onSaveUserInfo(user);
        this.notifyUserInfoChanged();
    }

    /**
     * 添加Session状态改变通知，在不用时候记得移除掉，避免内存泄漏
     */
    public void addSessionStateListener(@NonNull SessionStateListener listener) {
        if (mSessionStateListeners == null) {
            mSessionStateListeners = new ArrayList<>();
        }
        mSessionStateListeners.add(listener);
    }

    /**
     * 移除Session状态改变通知
     */
    public void removeSessionStateListener(@NonNull SessionStateListener listener) {
        if (mSessionStateListeners != null) {
            mSessionStateListeners.remove(listener);
        }
    }

    protected void notifyUserInfoChanged() {
        if (mSessionStateListeners == null) {
            return;
        }
        for (SessionStateListener listener : mSessionStateListeners) {
            listener.onUserInfoChanged(this);
        }
    }

}

File: kotlin/src/main/java/androidx/swift/session/SessionManager.kt
package androidx.swift.session

import android.content.Context

/**
 * Session管理器
 * @author RAE
 * @date 2022/06/12
 * @copyright Copyright (c) https://github.com/raedev All rights reserved.
 */
class SessionManager {

    companion object {
        /** 默认的Session提供者，请通过[Builder]进行初始化。 */
        lateinit var default: SessionDelegate
            private set
    }

    @Suppress("UNCHECKED_CAST")
    class Builder<T : Any>(private val context: Context, private val userClass: Class<T>) {

        private var isJsonProvider: Boolean = false
        private var enableEncrypt: Boolean = false
        private var sessionName: String? = null


        /**
         * 配置名称，默认为包名+session。
         * 当应用中有多个Session管理器的时候，使用SessionName进行区分。
         * @param name
         * @return
         */
        fun sessionName(name: String): Builder<T> {
            sessionName = name
            return this;
        }


        /**
         * 是否加密Session内容，防止root用户查看导致用户信息泄露
         * @param enableEncrypt
         * @return
         */
        fun enableEncrypt(enableEncrypt: Boolean): Builder<T> {
            this.enableEncrypt = enableEncrypt
            return this;
        }

        /**
         * Session管理器提供者使用JSON存储，默认是[SharedPreferencesSessionDelegate]
         * 除非一些特殊场景使用到（例如：Session信息保存到存储卡中）一般情况使用默认的即可。
         * @return
         */
        fun jsonProvider(): Builder<T> {
            this.isJsonProvider = true;
            return this;
        }

        /**
         *  构建默认的管理器并且对默认的[default]对象赋值。
         */
        fun build() {
            default = when (isJsonProvider) {
                true -> JsonSessionDelegate(context, sessionName, userClass)
                else -> SharedPreferencesSessionDelegate(context, sessionName, userClass)
            }
            val delegate = default as DefaultSessionDelegate<T>
            delegate.enableEncrypt = this.enableEncrypt
        }
    }

}